### PR TITLE
BUG: Fixed pd.unique on array of tuples

### LINKED ISSUE #16519: Regression from 0.19.2 to 0.20.1 in pandas.unique() when applied to list of tuples
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd

input = [(0, 0), (0, 1), (1, 0), (1, 1), (0, 0), (0, 1), (1, 0), (1, 1)]
print pd.unique(input)
```
#### Problem description
The code exits unexpectedly

```
Traceback (most recent call last):
  File "pandas_bug.py", line 6, in <module>
    pd.unique(input)
  File "/Users/johannes/.virtualenvs/pandas/lib/python2.7/site-packages/pandas/core/algorithms.py", line 351, in unique
    uniques = table.unique(values)
  File "pandas/_libs/hashtable_class_helper.pxi", line 1271, in pandas._libs.hashtable.PyObjectHashTable.unique (pandas/_libs/hashtable.c:21384)
ValueError: Buffer has wrong number of dimensions (expected 1, got 2)
```

#### Expected Output
The code works on pandas version 0.19.2 and produces the expected output
```
[(0, 0) (0, 1) (1, 0) (1, 1)]
```
Moreover this problem is not limited to MacOSX, but was also encounter on Ubuntu CI server.

#### Output of ``pd.show_versions()``

<details>
INSTALLED VERSIONS
------------------
commit: None
python: 2.7.10.final.0
python-bits: 64
OS: Darwin
OS-release: 15.6.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: None.None

pandas: 0.20.1
pytest: None
pip: 9.0.1
setuptools: 35.0.2
Cython: None
numpy: 1.12.1
scipy: None
xarray: None
IPython: None
sphinx: None
patsy: None
dateutil: 2.6.0
pytz: 2017.2
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: None
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: None
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: None
s3fs: None
pandas_gbq: None
pandas_datareader: None
None

</details>

### PR DESCRIPTION
**Pull request by TomAugspurger**
Closes #16519

### CODE REVIEW

File: doc/source/whatsnew/v0.20.2.txt [hunk 1]
```diff
@@ -39,7 +39,7 @@ Bug Fixes
 
 - Bug in using ``pathlib.Path`` or ``py.path.local`` objects with io functions (:issue:`16291`)
 - Bug in ``DataFrame.update()`` with ``overwrite=False`` and ``NaN values`` (:issue:`15593`)
-
+- Bug in :func:`pd.unique` on an array of tuples (:issue:`16519`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.20.2.txt]: I think has to be ``:func:`unique` ``?

File: pandas/tests/test_algos.py [hunk 2]
```diff
@@ -929,6 +929,22 @@ def test_unique_index(self):
             tm.assert_numpy_array_equal(case.duplicated(),
                                         np.array([False, False, False]))
 
+    @pytest.mark.parametrize('arr, unique', [
+        ([(0, 0), (0, 1), (1, 0), (1, 1), (0, 0), (0, 1), (1, 0), (1, 1)],
```
Comment (jreback) [inline pandas/tests/test_algos.py]: also add an example of this in ``pd.unique`` itself.

**Review by jreback — APPROVED** [review]
lgtm.

**Comment by chris-b1** [comment]
I think you could back out this change from #16434

https://github.com/TomAugspurger/pandas/blob/35da5b9e60993b722760d52dc8dcc0a14129748e/pandas/core/algorithms.py#L391

**Comment by jreback** [comment]
I with agree @chris-b1 comment, yes that looks right.

File: doc/source/whatsnew/v0.20.2.txt (hunk 1 above)
Comment (TomAugspurger) [inline doc/source/whatsnew/v0.20.2.txt]: You're correct.

**Comment by TomAugspurger** [comment]
> I think you could back out this change from #16434

The regression test from #16434 fails if I revert the change. The difference being

```python
(Pdb++) values  # this is with the fix reverted
array([[1, 'a']], dtype=object)
(Pdb++) lib.list_to_object_array(list([(1, 'a')]))  # this is the fix from 16434
array([(1, 'a')], dtype=object)
```

So an array of lists vs. an array of tuples. Is that correct?

**Comment by jreback** [comment]
you may just need to call
_ensure_arraylike there instead of the isinstance check

**Comment by TomAugspurger** [comment]
Using `_ensure_arraylike` failed on an empty array `pd.Series([1, 2]).isin([])` since it's a float instead of object dtype. I can handle that case if you want, or just leave the isinstance checks. Not sure which is cleaner really.

**Comment by jreback** [comment]
@TomAugspurger added a commit. should fix up I think.

**Comment by jreback** [comment]
@TomAugspurger looks this broke it. ok just revert my commit and merge your changes. This is a very touchy area. I we are doing the right things in the tests, but just tricky to get exactly right.

**Comment by jreback** [comment]
@TomAugspurger rebased to remove my commit.

**Comment by jreback** [comment]
thanks!

**Pull request merged**